qcfractal/storage_sockets/sqlalchemy_socket.py

```diff
@@ -1135,33 +1135,80 @@ def add_results(self, record_list: List[ResultRecord]):
             Dict with keys: data, meta
             Data is the ids of the inserted/updated/existing docs
         """
-
         meta = add_metadata_template()
 
-        result_ids = []
-        with self.session_scope() as session:
-            for result in record_list:
+        result_ids = ["placeholder"] * len(record_list)
+
+        results_list = []
+        existing_res = {}
+        new_record_idx, duplicates_idx = [], []
+        conds = []
+
+        conds = [
+            and_(
+                ResultORM.program == res.program,
+                ResultORM.driver == res.driver,
+                ResultORM.method == res.method,
+                ResultORM.basis == res.basis,
+                ResultORM.keywords == res.keywords,
+                ResultORM.molecule == res.molecule,
+            )
+            for res in results_list
+        ]
 
-                doc = session.query(ResultORM).filter_by(
-                    program=result.program,
-                    driver=result.driver,
-                    method=result.method,
-                    basis=result.basis,
-                    keywords=result.keywords,
-                    molecule=result.molecule,
+        with self.session_scope() as session:
+            docs = session.query(ResultORM).filter(or_(*conds)).all()
+
+            existing_res = {
+                (doc.program, doc.driver, doc.method, doc.basis, str(doc.keywords), str(doc.molecule)): doc
```

Comment:
I am not sure why you are casting `keywords` here, None is perfectly valid in dictionary tuple keys.

Reply:
as we're looking through a found_items dictionary, the keyword item may return a valid object(instead of ```None```), which is of type ```KeywordSet```, so instead of instantiating a ```KeywordSet``` object, I represent it with the given input string and compare with that.

Reply:
This should not be the case snipping from the ORM:
```
    keywords = Column(Integer, ForeignKey("keywords.id"))
    keywords_obj = relationship(KeywordsORM, lazy="select")
```

Plus if the keywords field is a `KeywordSet` casting to a string doesn't give you want you want. The string representation is not guaranteed to be unique.